contracts/serverless-v2/Executors.sol

```diff
@@ -0,0 +1,527 @@
+            address enclaveAddress = selectedNodes[index];
+            executors[enclaveAddress].activeJobs += 1;
+
+            // if jobCapacity reached then delete from the tree so as to not consider this node in new jobs allocation
+            if(executors[enclaveAddress].activeJobs == executors[enclaveAddress].jobCapacity)
+                _deleteIfPresent(enclaveAddress);
+        }
+    }
+
+    function _selectNodes(
+        uint256 _noOfNodesToSelect
+    ) internal view returns (address[] memory selectedNodes) {
+        uint256 randomizer = uint256(keccak256(abi.encode(blockhash(block.number - 1), block.timestamp)));
+        selectedNodes = _selectN(randomizer, _noOfNodesToSelect);
+        // require(selectedNodes.length != 0, "NO_EXECUTOR_SELECTED");
+    }
+
+    function _releaseExecutor(
+        address _enclaveAddress
+    ) internal {
+        if(!executors[_enclaveAddress].draining) {
+            // node might have been deleted due to max job capacity reached
+            // if stakes are greater than minStakes then update the stakes for executors in tree if it already exists else add with latest stake
+            if(executors[_enclaveAddress].stakeAmount >= MIN_STAKE_AMOUNT)
+                _upsert(_enclaveAddress, uint64(executors[_enclaveAddress].stakeAmount / STAKE_ADJUSTMENT_FACTOR));
+            // remove node from tree if stake falls below min level
+            else
+                _deleteIfPresent(_enclaveAddress);
+        }
+
+        executors[_enclaveAddress].activeJobs -= 1;
+    }
+
+    function _slashExecutor(
+        address _enclaveAddress,
+        bool _isNoOutputSubmitted,
+        address _jobOwner
+    ) internal returns (uint256) {
+        uint256 totalComp = executors[_enclaveAddress].stakeAmount * SLASH_PERCENT_IN_BIPS / SLASH_MAX_BIPS;
+        executors[_enclaveAddress].stakeAmount -= totalComp;
+
+        if(_isNoOutputSubmitted) {
+            // transfer the slashed comp to job owner
+            TOKEN.safeTransfer(_jobOwner, totalComp);
+        }
+        else {
+            // transfer the slashed comp to payment pool
+            TOKEN.safeTransfer(paymentPool, totalComp);
+        }
+
+        _releaseExecutor(_enclaveAddress);
+        return totalComp;
+    }
+
+    //-------------------------------- internal functions end ----------------------------------//
+
+    //-------------------------------- external functions start ----------------------------------//
+
+    function selectExecutors(
+        uint256 _noOfNodesToSelect
+    ) external onlyRole(JOBS_ROLE) returns (address[] memory selectedNodes) {
+        return _selectExecutors(_noOfNodesToSelect);
+    }
+
+    function releaseExecutor(
+        address _enclaveAddress
+    ) external onlyRole(JOBS_ROLE) {
+        _releaseExecutor(_enclaveAddress);
+    }
+
+    function slashExecutor(
+        address _enclaveAddress,
+        bool _isNoOutputSubmitted,
+        address _jobOwner
```

Comment:
should this just be combined to have one single recipient address? The jobs contract can decide between owner vs pool.